scan/scan_common.go

```diff
@@ -163,57 +170,72 @@ func (fs FamilySet) RunScans(host, ip, family, scanner string, dur time.Duration
 		return nil, err
 	}
 
-	results := make(chan map[string]FamilyResult)
-	timeout := make(chan bool)
-
-	if dur > 0 {
-		go startTimer(dur, timeout)
-	}
+	results := make(chan *Result)
 
-	familyResults := make(map[string]FamilyResult)
+	var familyWG sync.WaitGroup
+	familyWG.Add(len(fs))
+	go func() {
+		familyWG.Wait()
+		close(results)
+	}()
 
 	go func() {
 		for familyName, family := range fs {
+			var scannerWG sync.WaitGroup
 			if familyRegexp.MatchString(familyName) {
-				scannerResults := make(map[string]ScannerResult)
+				scannerWG.Add(len(family.Scanners))
 				for scannerName, scanner := range family.Scanners {
-					if scannerRegexp.MatchString(scannerName) {
-						grade, output, err := scanner.Scan(addr, hostname)
-						if grade != Skipped {
-							result := ScannerResult{
-								Grade:  grade.String(),
-								Output: output,
-							}
+					go func(scannerName string, scanner *Scanner) {
+						if scannerRegexp.MatchString(scannerName) {
+							grade, output, err := scanner.Scan(addr, hostname)
+							var errStr string
 							if err != nil {
-								result.Error = err.Error()
+								errStr = err.Error()
+							}
+							results <- &Result{
+								Family:  familyName,
+								Scanner: scannerName,
+								ScannerResult: ScannerResult{
+									Grade:  grade.String(),
+									Output: output,
+									Error:  errStr,
+								},
 							}
-
-							scannerResults[scannerName] = result
 						}
-					}
-				}
-
-				if len(scannerResults) > 0 {
-					familyResults[familyName] = scannerResults
+						scannerWG.Done()
+					}(scannerName, scanner)
 				}
 			}
+			go func() {
+				scannerWG.Wait()
+				familyWG.Done()
+			}()
 		}
-		results <- familyResults
 	}()
-
-	select {
-	case <-timeout:
-		log.Warningf("scan: %s timed out after %v", host, dur)
-		return familyResults, nil
-	case res := <-results:
-		return res, nil
-	}
+	return results, nil
 }
 
-func startTimer(t time.Duration, timeout chan bool) {
-	time.AfterFunc(t, func() {
-		timeout <- true
-	})
+// ProcessResults converts a channel of results into a JSON marshallable map of results.
+func ProcessResults(resultChan <-chan *Result, timeout time.Duration) (results map[string]FamilyResult) {
+	results = make(map[string]FamilyResult)
+	end := time.After(timeout)
+	for {
+		select {
+		case result := <-resultChan:
+			if result == nil {
+				return
+			}
+
+			if results[result.Family] == nil {
+				results[result.Family] = make(FamilyResult)
+			}
+
+			results[result.Family][result.Scanner] = result.ScannerResult
+
+		case <-end:
+			return
```

Comment:
There's a bit of a hidden bug here. We can't inform the running scanners that we've timed out, so when they will continue to run and block indefinitely on sending to the `results` channel.

(We could close results here and defer/recover on the `send on closed channel` panic inside `RunScans`)